---
title: "RiboSeek Anlaysis"
author: "SemiQuant"
date: "2/28/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r packages, include=FALSE, eval=FALSE}
install.packages(c("plotly", "tidyverse", "BiocManager", "DESeq2", "parallel", "xtail", "devtools",
  "ggplot2", "heatmaply", "adegenet"), quietly = T)
BiocManager::install(c("riboSeqR", "systemPipeR", "RiboDiPA"))
devtools::install_github("xryanglab/xtail", dependencies = TRUE)
# install.packages("xtail",
#                  repos = NULL, 
#                  type = "source")
devtools::install_github("LabTranslationalArchitectomics/riboWaltz", dependencies = TRUE, 
               build_opts = c("--no-resave-data"), build_vignettes = TRUE)
```

```{r load packages, message=FALSE, warning=FALSE, include=FALSE}
require(plotly)
require(tidyverse)
require(BiocManager)
require(DESeq2)
require(parallel)
require(xtail)
require(devtools)
require(ggplot2)
require(heatmaply)
require(adegenet)
require(riboSeqR)
require(systemPipeR)
require(RiboDiPA)
require(riboWaltz)
require(xtail)
```

## Inspiration

-   <https://bioconductor.org/packages/release/bioc/html/riboSeqR.html>

-   <https://github.com/xryanglab/xtail>

-   <https://bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeRIBOseq.html>

-   <https://github.com/LabTranslationalArchitectomics/riboWaltz>

-   <https://github.com/ratschlab/ribodiff>

-   <https://pubmed.ncbi.nlm.nih.gov/19213877/#&gid=article-figures&pid=fig-4-uid-3>

-   <https://bioconductor.org/packages/devel/bioc/vignettes/RiboDiPA/inst/doc/RiboDiPA.html>

## Read in processed data and metadata, and check/clean

```{r data}

```

### Deseq

```{r}
files_in <- list.files("/Users/SemiQuant/Downloads/yux", "*.NC_000962.HTSeq.counts",
                       full.names = T, recursive = T)

genes <- read_tsv(files_in[1], col_names = F, show_col_types = F, col_select = "X1")

counts <- files_in %>%
  map(read_tsv, col_names = F, show_col_types = F, col_select = "X2") %>%
  purrr::reduce(cbind)

colnames(counts) <- gsub(".NC_000962.HTSeq.counts", "", basename(files_in))
counts <- cbind(genes, counts)

counts <- counts %>% 
  dplyr::slice(1:(nrow(genes)-5))

counts %>%
  dplyr::select(-1) %>% 
  mutate(n = 1:nrow(counts)) %>%
  pivot_longer(!n, names_to = "Gene", values_to = "count") %>%
  plot_ly(x = ~n, y = ~count, type = 'scatter', mode = 'lines', color = ~Gene)


# filter those without any counts but do it after selecting groups etc
counts <- counts[rowSums(counts[-c(1)], na.rm = T) >= 10, ]

gen_names <- counts$X1
rownames(counts) <- gen_names

counts <- counts %>%
  dplyr::select(-c(1))
genes <- NULL




meta <- data.frame(phen = c("w", "w", "w", "m", "m", "m"))

dds_1 <- DESeqDataSetFromMatrix(countData = counts,
                                colData = meta,
                                # tidy = T,
                                design = ~ phen) #Cell + TBstatus
dds_1 <- estimateSizeFactors(dds_1)
dds_1 <- DESeq(dds_1, parallel = T)
res_1 <- results(dds_1)

vst_1_blind <- vst(dds_1, blind = T)
vst_1_blind_cor <- cor(assay(vst_1_blind))



# Heatmap of all transcripts color by groups ------------------------------------------------
hmap_QC_1 <- heatmaply(vst_1_blind_cor,
                       # symm = T,
                       col_side_colors = meta$phen,
                       row_side_colors = meta$phen,
                       hclust_method = "complete",
                       dist_method = "manhattan",
                       fontsize_row = 5,
                       fontsize_col = 5,
                       # row_text_angle = 45,
                       plot_method = "plotly"
) 



PCA <- dudi.pca(df = t(assay(vst_1_blind)), center = T,
                scale = T)

p1 <- qplot(data = PCA$l1, x = RS1, y = RS2, colour = fac2col(meta$phen))
p1 <- p1 + stat_ellipse(geom = "polygon", 
                        level = 0.95,
                        type = "norm",
                        alpha = 0.3, aes(fill = meta$phen)) +
  stat_ellipse(type = "euclid", linetype = 2, alpha = 0.3) # level = 0.2)


pca_sum <- factoextra::get_eigenvalue(PCA)

p1 <- ggplotly(p1 +
                 # labs(title = title) + #, subtitle = "A subtitle"
                 xlab(label = paste("PC1", "(", round(pca_sum$variance.percent[1], 0), "%)")) +
                 ylab(label =  paste("PC2", "(", round(pca_sum$variance.percent[2], 0), "%)")
                 )
) %>% 
  layout(
    legend = list(
      orientation = 'h',
      y=-0.15)
  )
p1 



# Corr heatmap ---------------
(hp_corr <- heatmaply(vst_1_blind_cor,
                      symm = T,
                      col_side_colors = meta$phen,
                      row_side_colors = meta$phen,
                      hclust_method = "complete",
                      dist_method = "euclidean", #"manhattan",
                      fontsize_row = 5,
                      fontsize_col = 5,
                      # k_col=9,
                      # k_row = 4,
                      dendrogram = "both",
                      # row_text_angle = 45,
                      plot_method = "plotly"
))



# save(dds_in, file = "dds_in.Rdat")


vol_plot <- function(deSeq_results = NULL, adj_p = 0.05, num_annotate = 10, effect_sizeThresh = 2, makePlot = T){
  volc.plot <- data.frame(ID = rownames(deSeq_results),
                          P = deSeq_results$pvalue,
                          P.adj = deSeq_results$padj,
                          EffectSize = deSeq_results$log2FoldChange,
                          threshold = as.factor(deSeq_results$padj <= adj_p)
                          # comparison = c("TB vs No TB")
  )
  
  volc.plot$threshold <- as.factor(volc.plot$P.adj <= adj_p & abs(volc.plot$EffectSize) >= effect_sizeThresh)
  
  
  # rewrite this, in a rush
  effect_sizeThresh_2 <- abs(volc.plot[volc.plot$P.adj <= adj_p,]$EffectSize)
  effect_sizeThresh_2 <- effect_sizeThresh_2[order(abs(effect_sizeThresh_2), decreasing = T)][num_annotate]
  
  
  volc.plot$lab <- ifelse(volc.plot$P.adj <= adj_p & abs(volc.plot$EffectSize) >= effect_sizeThresh_2, as.character(volc.plot$ID), "")
  
  volc_DE_1 <- ggplot2::ggplot(data = volc.plot,
                               ggplot2::aes(x = EffectSize, y = -log10(P),
                                            colour = threshold, text = ID)) +
    ggplot2::geom_point(alpha = 0.4, size = 1.75) +
    # xlim(c(-6, 6)) +
    ggplot2::xlab("Effect Size") + ggplot2::ylab("-log10 p-value") +
    ggplot2::theme_bw() +
    ggplot2::theme(legend.position = "none") +
    ggplot2::theme(panel.grid.major = ggplot2::element_blank(),
                   panel.grid.minor = ggplot2::element_blank()) +
    ggplot2::geom_text(ggplot2::aes(label = volc.plot$lab),
                       hjust = 0, vjust = 0, colour = "dark grey", size = 2, nudge_x = 0.2)
  
  volc_DE_1 <- volc_DE_1 + ggplot2::geom_hline(yintercept = -log10(min(volc.plot$P[volc.plot$P.adj <= adj_p])), colour = "grey", linetype = "longdash") +
    ggplot2::geom_vline(xintercept = c(-1*effect_sizeThresh, effect_sizeThresh), colour = "grey",linetype = "longdash")
  # ggplot2::geom_hline(yintercept = -log10(adj_p), colour = "grey", linetype = "longdash") +
  
  
  if (makePlot)
    return(volc_DE_1)
  else
    return(volc.plot)
  
}

vol_plot(deSeq_results = res_1, adj_p = 0.05, num_annotate = 10, effect_sizeThresh = 2, makePlot = T)

# 
# 
# heatmaply(assay(vst_1_blind)[1:10,],
#           # col_side_colors = colnames(vst_1_blind),
#           # row_side_colors = meta$phen,
#           hclust_method = "complete",
#           dist_method = "manhattan",
#           fontsize_row = 5,
#           fontsize_col = 5,
#           # row_text_angle = 45,
#           plot_method = "plotly"
# )


```

```{r systemPipeR}
require(systemPipeR)
require(GenomicFeatures)
require(ggplot2)
require(grid)

gtf_file <- "/Users/SemiQuant/Bioinformatics/Projects/riri/references/NC_000962.gff"
txdb <- makeTxDbFromGFF(file = gtf_file)
feat <- genFeatures(txdb, featuretype = "all", reduce_ranges = TRUE,
                    upstream = 50, downstream = 50, verbose = TRUE)



bam_path <- "/Users/SemiQuant/Downloads/riboDelete/riboSeq_R1.bam"
(bf <- BamFileList(bam_path))

fc <- featuretypeCounts(bfl = bf,
                        grl = feat, singleEnd = TRUE, readlength = NULL, type = "data.frame")


p <- plotfeaturetypeCounts(x = fc, graphicsfile = "~/Downloads/riboDelete/featureCounts.png",
                           graphicsformat = "png", scales = "fixed", anyreadlength = TRUE,
                           scale_length_val = NULL)

# feat2 <- cdsBy(txdb, "tx")
# 
# fcov <- featureCoverage(bfl = bf, grl = feat2,
#                         resizereads = NULL, readlengthrange = NULL, Nbins = 20, method = mean,
#                         fixedmatrix = TRUE, resizefeatures = TRUE, upstream = 20,
#                         downstream = 20, 
#                         outfile = "~/Downloads/riboDelete/featureCoverage.xls",
#                         overwrite = TRUE)
# 
# plotfeatureCoverage(covMA = fcov, method = mean, scales = "fixed",
#                     extendylim = 2, scale_count_val = 10^6)
# 
# # nucleotide level coverage along entire transcripts/CDSs
# cov <- featureCoverage(bfl = BamFileList(outpaths[1:2]), grl = grl[1],
#                        resizereads = NULL, readlengthrange = NULL, Nbins = NULL,
#                        method = mean, fixedmatrix = FALSE, resizefeatures = TRUE,
#                        upstream = 20, downstream = 20, outfile = NULL)

```

```{r psite}
require(RiboDiPA)
bam_path <- list.files(pattern = ".bam$", "/Users/SemiQuant/Downloads/riboDelete/riboSeq_R1_riri", full.names = T)
gtf <- "/Users/SemiQuant/Bioinformatics/Projects/riri/references/NC_000962.gff"

classlabel <- data.frame(condition = c("mutant",  "wildtype"), comparison = c(2, 1))

# https://bioconductor.org/packages/devel/bioc/vignettes/RiboDiPA/inst/doc/RiboDiPA.html#p-site-mapping
## Perform individual P-site mapping procedure
data.psite <- psiteMapping(bam_file_list = bam_path, 
                           gtf_file = gtf, psite.mapping = "auto", cores = 5)

## P-site mapping offset rule generated
data.psite$psite.mapping

# ## Use user specified psite mapping offset rule
# offsets <- cbind(qwidth = c(28, 29, 30, 31, 32), 
#                  psite = c(18, 18, 18, 19, 19))
# data.psite2 <- psiteMapping(bam_path[1:4], bam_path[5], 
#                             psite.mapping = offsets, cores = 2)


## Merge the P-site data into bins with a fixed or an adaptive width
data.binned <- dataBinning(data = data.psite$coverage, bin.width = 0, 
                           zero.omit = FALSE, bin.from.5UTR = TRUE, cores = 2)

## Merge the P-site data on each codon
data.codon <- dataBinning(data = data.psite$coverage, bin.width = 1, 
                          zero.omit = FALSE, bin.from.5UTR = TRUE, cores = 2)

## Merge the P-site data on each exon and perform differential pattern analysis
result.exon <- diffPatternTestExon(psitemap = data.psite, 
                                   classlabel = classlabel, method = c('gtxr', 'qvalue'))

## Perform differential pattern analysis
result.pst <- diffPatternTest(data = data.binned, 
                              classlabel = classlabel, method=c('gtxr', 'qvalue'))


## Plot ribosome per nucleotide tracks of specified genes.
plotTrack(data = data.psite, genes.list = c("YDR050C", "YDR064W"),
          replicates = NULL, exons = FALSE)

## Plot binned ribosome tracks of siginificant genes: YDR086C and YDR210W.
## you can specify the thrshold to redefine the significant level
plotTest(result = result.pst, genes.list = NULL, threshold = 0.05) 

```

```{r count_metagene_profile}

nme="test"

f <- "/Users/SemiQuant/Downloads/riboDelete/test_count_metagene_profile.txt"
cnt <- read_tsv(f, comment = "#")
cnt$Sample <- "S1"

cnt2 <- cnt
cnt2$Sample <- "S2"
cnt <- rbind(cnt, cnt2)




# do this for all samples together
cnt %>% 
  filter(!is.na(metagene_average)) %>% 
  plot_ly(x = ~x, y = ~metagene_average, type = 'scatter', mode = 'lines', color = ~Sample)

cnt %>% 
  filter(!is.na(metagene_average)) %>% 
  dplyr::select(-regions_counted) %>% 
  pivot_wider(names_from = Sample,
              values_from = metagene_average
              ) %>% 
  dplyr::select(-x) %>% 
heatmaply(#row_side_colors = .colnames(),
          hclust_method = "complete",
          dist_method = "manhattan",
          fontsize_row = 5,
          fontsize_col = 5,
          # row_text_angle = 45,
          plot_method = "plotly"
) 

```

<!-- Not working with current alignment method -->

```{r}
require(tidyverse)
require(riboWaltz)
# browseVignettes("riboWaltz")

gtf_file <- "/Users/SemiQuant/Bioinformatics/Projects/riri/references/NC_000962.gtf_5p.gtf"
bam_path <- "/Users/SemiQuant/Downloads/riboDelete"


annotation_dt <- create_annotation(gtfpath = gtf_file)

# to get the other names
gtf <- rtracklayer::import(gtf_file)
gtf <- as.data.frame(gtf)
annots <- gtf %>% 
  select(locus_tag, transcript_id) %>% 
  unique
annots <- annots[match(annotation_dt$transcript, annots$transcript_id),]
table(annotation_dt$transcript == annots$transcript_id)
# annotation_dt$transcript <- annots$transcript_id

# make this up
annotation_dt$l_utr5 <- annotation_dt$l_utr3 <- 30




reads_list <- bamtolist(bamfolder = bam_path, annotation = annotation_dt, transcript_align = F)
filtered_list <- duplicates_filter(data = example_reads_list,
                                   extremity = "both")

filtered_list <- length_filter(data = example_reads_list,
                               length_filter_mode = "periodicity",
                               periodicity_threshold = 70)


psite_offset <- psite(reads_list, flanking = 6, extremity = "auto")






bamtolist

```
